app/api/internal/applications.rb

```diff
@@ -2,6 +2,84 @@ module ThreeScale
   module Backend
     module API
       internal_api '/services/:service_id/applications' do
+        # Helper method to save a single application with all its associated data
+        def save_application_with_data(service_id, app_data)
+          app_id = app_data[:id]
```

Comment:
We need to validate the presence of the `id` field. 
Currently, this is considered successful:
```
{
  "applications": [
    {
      "state": "ok"
    }
  ]
}
```

and it creates invalid entries in redis:
```
[db8] > keys *
1) "application/service_id:2/id:/state"
2) "service_id:2/applications"

[db8] > get application/service_id:2/id:/state
"ok"

[db8] > smembers service_id:2/applications
1) ""
```

Reply:
Ah, right, at first I thought this validation would only apply for the new batch updates - then it wouldn't be concerning. But indeed, it can happen for any application updates, and indeed, we don't know what's there in production.

So, I'd remove the `ApplicationStateInvalid` error.

Indeed, in my example, I used `state=ok`, but I guess I wasn't clear enough that that was not bothering me, the only problem I saw was a missing ID, not the "invalid" state.